presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,762 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.type;
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.airlift.slice.XxHash64;
+import sun.misc.Unsafe;
+
+import java.lang.reflect.Field;
+import java.math.BigInteger;
+import java.nio.ByteOrder;
+
+import static com.facebook.presto.spi.type.Decimals.MAX_PRECISION;
+import static com.facebook.presto.spi.type.Decimals.longTenToNth;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
+
+/**
+ * 128 bit unscaled decimal arithmetic.
+ * <p>
+ * Based on: https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/type/UnsignedInt128.java
+ * and https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/type/Decimal128.java
+ */
+public final class UnscaledDecimal128Arithmetic
+{
+    public static final int UNSCALED_DECIMAL_128_SLICE_LENGTH = 2 * SIZE_OF_LONG;
```

Comment:
I have to move private `NUMBER_OF_LONGS` above `UNSCALED_DECIMAL_128_SLICE_LENGTH`. Otherwise it would be illegal foward reference. I don't like privates above public fields, but I can move it there.